Refactor and improve make_fx testing

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #103198
* #103197
* __->__ #103196

This is in preparation for the custom_op_compile_check utility, which
will call the newly refactored function.

This PR:
- splits off code into helper functions
- adds clearer error messages
- stops updating the inputs destructively (leading to slightly slower
tests)

zou3519 committed:
Refactor and improve make_fx testing

This is in preparation for the custom_op_compile_check utility, which
will call the newly refactored function.

This PR:
- splits off code into helper functions
- adds clearer error messages
- stops updating the inputs destructively (leading to slightly slower
tests)

[ghstack-poisoned]

zou3519 committed:
Update on "Refactor and improve make_fx testing"


This is in preparation for the custom_op_compile_check utility, which
will call the newly refactored function.

This PR:
- splits off code into helper functions
- adds clearer error messages
- stops updating the inputs destructively (leading to slightly slower
tests)

[ghstack-poisoned]

zou3519 committed:
Update on "Refactor and improve make_fx testing"


This is in preparation for the custom_op_compile_check utility, which
will call the newly refactored function.

This PR:
- splits off code into helper functions
- adds clearer error messages
- stops updating the inputs destructively (leading to slightly slower
tests)

[ghstack-poisoned]

zou3519 committed:
Update on "Refactor and improve make_fx testing"


This is in preparation for the custom_op_compile_check utility, which
will call the newly refactored function.

This PR:
- splits off code into helper functions
- adds clearer error messages
- stops updating the inputs destructively (leading to slightly slower
tests)

[ghstack-poisoned]

bdhirsh commented on torch/testing/_internal/optests/make_fx.py:
```diff
@@ -0,0 +1,87 @@
+        "op(*args, **kwargs) and make_fx(op)(*args, **kwargs) produced different "
+        "values. This could mean that your abstract impls (meta/FakeTensor impls) "
+        "are incorrect, that your operator is not completely traceable (e.g., "
+        "it relies on some global state), or that there is a bug in make_fx."
```
is it worth including the traced graph in the error message for sanity?

zou3519 replied:
we could!

zou3519 replied:
On second thought, I don't want to, because it means the error message might get long, especially if you have a CompositeImplicitAutograd op that decomposes into something like matmul. It's pretty easy for the user to do the make_fx(op)(*args, kwargs) themselves and see what it traces out as.

zou3519 commented on test/test_proxy_tensor.py:
```diff
@@ -1566,8 +1567,6 @@ def f(a, b, c, d, e):
     # decomp not implemented
     xfail('unique', ''),
-    # in-place has a different signature than out-of-place
-    xfail('uniform', ''),
```
Previously, we were setting the random seed for trace_f(*args) but not setting it for f(*args). This PR changes it so that we set the random seed for both. Because random operators are run in the same order, we can expect the output of both of these to be the same.

bdhirsh replied:
oof, nice catch

soulitzer commented on torch/testing/_internal/optests/make_fx.py:
```diff
@@ -0,0 +1,87 @@
+        if not x.dtype.is_floating_point:
+            return x
+        with torch.no_grad():
+            y = x.detach().clone().uniform_(0, 1)
```
nit: you can do this in one line :)
`return x.detach().clone().uniform_(0, 1).requires_grad_(x.requires_grad)`

(no need to have both no-grad and detach)

zou3519 replied:
Good point!

zou3519 committed:
Update on "Refactor and improve make_fx testing"


This is in preparation for the custom_op_compile_check utility, which
will call the newly refactored function.

This PR:
- splits off code into helper functions
- adds clearer error messages
- stops updating the inputs destructively (leading to slightly slower
tests)

[ghstack-poisoned]
